### DISCUSSION TITLE
State Root Reasoning on L1

### DISCUSSION THREAD

**Discussion by novaknole** [ethereum-optimism/developers#305] [Core Protocol]
### Did you check the documentation?

- [X] Yes

### Did you read the specs?

- [X] Yes
- [ ] No

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

My question will be related to optimism v0 that you guys were building 3 years ago. Currently, it mightn't be the same as my question will relate to, but I'd appreciate the answer as if you were answering optimism v0(ofc, if you know, if not, then you can answer it with the latest optimism in mind)

As we know, sequencer publishes calldatas and state roots to L1. The reason it publishes calldatas is that it should be super easy to recover L2 chain in case something goes wrong. Since ethereum is secure, we rely on it so that in case Optimism goes down, we can still recover the chain by re-executing calldatas.

Now, I wonder why state roots are published. One reason you will say is that it allows fraud proofs, such that verifier can check if the state root published is correct and if not, prove on L1 why it's not correct, but I believe this is not the reason why we need state roots published on L1. Even if you don't publish it, L2 nodes still exist, they still wait for incoming blocks from sequencer and in case, tx is not valid in a block, L2 nodes won't add it to their chain. Everything will still be great.

So, then I researched and read that, Its purpose is to allow secure withdrawals.

> Posting state roots on L1 is a key characteristic of rollups that allows for withdrawals. If a rollup does not post state roots on L1, it falls short of a fundamental component of a bridged rollup.

I am curious how this works. When you want to deposit on L2, you call L1's smart contract's `deposit` function which emits the event to which L2 node listens to and runs it on L2's EVM.  All clear now, but it seems like state roots on L1 is needed to withdraw funds from L2 to L1. 

1. I would appreciate the reasoning why. What wouldn't work exactly if state roots weren't on L1.
2. I have been looking into https://github.com/ethereum-optimism/contracts/tree/df10de974629615b8597ed9bdbcd1b5ec2c0c9b7 but couldn't find where withdraw exists such that it uses state root check validation.

Thank you so much.

**Comment by hash1go** [comment]
> I would appreciate the reasoning why. What wouldn't work exactly if state roots weren't on L1.

Regarding the necessity of posting state roots on L1, it's crucial for ensuring that the state transitions on the OP Mainnet occur correctly according to the EVM. While Ethereum guarantees the correctness of the world state transitions through its consensus algorithm, Gasper, OP Mainnet being a rollup solution relies on Ethereum to provide this assurance.

If state roots weren't submitted to L1, there wouldn't be a mechanism to ensure that the state transitions on OP Mainnet are properly validated according to the EVM rules.

> I have been looking into https://github.com/ethereum-optimism/contracts/tree/df10de974629615b8597ed9bdbcd1b5ec2c0c9b7 but couldn't find where withdraw exists such that it uses state root check validation.

You can visit:
https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/withdrawals.md

**Reply by novaknole** [reply]
> If state roots weren't submitted to L1, there wouldn't be a mechanism to ensure that the state transitions on OP Mainnet are properly validated according to the EVM rules.

As far as I know and as we talked about on the previous topic, L2 nodes receive blocks from sequencer and they only add it to their chain if the txs inside the block are valid. So they validate the block header and txs inside it. This results in correct L2 nodes. To me, this seems to work without even state roots published on L1.  In this circumstance/scenario, what could go wrong or what am I missing ?

**Reply by hash1go** [reply]
Consider a scenario where the entity responsible for submitting state roots proposes a state root that is entirely different from the block generated in L2. Without a mechanism within the OP Mainnet to reach consensus on the submitted state roots, it is theoretically possible for this to happen.

**Reply by novaknole** [reply]
While I get your point, what I don't get is why my mechanism isn't enough. By my mechanism, i simply mean that sequencer never needs to publish state roots on L1, and it simply shares blocks to L2 nodes. L2 nodes will reject the block with invalid txs as they loop through the L2 block's txs, execute them in their own EVM and will quickly realize that tx is invalid. Makes sense ?

**Reply by hash1go** [reply]
Indeed, if an individual sets up a node with the capability to execute state transitions according to the EVM specifications, they can believe that their node is correct. However, consider a scenario where a majority of network participants collude to produce false state transitions. In such a case, external entities have no way to determine which state transitions are legitimate.

This issue is addressed in Bitcoin (although not EVM-based) through Nakamoto consensus, and in Ethereum through the Gasper consensus mechanism. Additionally, it's important to recognize that the primary method to verify the correctness of state transitions is, fundamentally, to execute the transactions.

**Reply by novaknole** [reply]
In bitcoin, it's addressed by proof of work, but if majority collude maliciously, you got the same problem in bitcoin.

As far as I understand, you base your logic on a fact that majority nodes wouldn't collude on bitcoin because it already has so many nodes, whereas on OP mainnet, it's easier to collude as there won't be those many nodes. 

**Q1:** Is this what you meant ?

If so, OP Mainnet's whole idea is to not depend on L2 nodes. Basically, If I want to know the balance of my account, first, I ask L2 node to give me the balance and the state trie proof and I use the root hash from L1 chain for that tree to verify. This ensures that even if ALL L2 nodes are malicious, that won't make anything horrible/broken as long as there's a single verifier existing.

**Q2:** All correct ?

**Reply by hash1go** [reply]
In the context of Bitcoin implementing Nakamoto consensus, external entities can trust the network as a whole if they believe that over 51% of it is honest. However, without consensus mechanism, external parties have no way to confirm if state transitions are correct unless they participate in the network by running a full node.

Your method implies reliance on a single verifier. Indeed, in a situation where a specific individual maintains the entire infrastructure of an Optimistic rollup and is the sole user of that rollup, there might be no need to submit state roots to L1. This is because the trust and verification would be concentrated in the hands of that single participant.

**Reply by novaknole** [reply]
Yes, I think I was correct in my last reply. In bitcoin, if 51% is honest, then at some point, longest chain rule will apply and make the state of the network correct.

In the OP mainnet case, we dont have longest chain rule and we should design the system such that even if 99% of L2 nodes are malicious, our system still should succeed. Even if only 1 verifier out of the whole OP mainnet nodes is honest, with L1 roots, system still work. 

Though, when a user(external entity) asks for a balance to L2 node, that user cant trust it directly the same way It would on L1. So I think the flow is this: user would get the latest state root from L1 and then ask the L2 node for state proof and then use that state ROOT to verify the proof. 

I think thats my final observation and I am sure you exactly agree with that, just because of my non-full trust in myself, I will ask again: am I correct ? 😂

**Reply by hash1go** [reply]
> In the OP mainnet case, we dont have longest chain rule and we should design the system such that even if 99% of L2 nodes are malicious, our system still should succeed. Even if only 1 verifier out of the whole OP mainnet nodes is honest, with L1 roots, system still work.

In such a case, while I believe the network would become practically unusable, the system would still function as designed. Discussing what happens in this scenario honestly goes beyond my understanding. You might find this blog post helpful: https://vitalik.eth.limo/general/2020/08/20/trust.html

> Though, when a user(external entity) asks for a balance to L2 node, that user cant trust it directly the same way It would on L1. So I think the flow is this: user would get the latest state root from L1 and then ask the L2 node for state proof and then use that state ROOT to verify the proof.

Strictly speaking, you would indeed need to use the state root from L1 for verification. However, in reality, users who do not operate a full node are essentially placing their trust in a single full node, whether it's on L1 or L2.

**Reply by novaknole** [reply]
Sure... L1 state roots seem to exist for the following: "let it be here, we won't need it, but in case there's a need, we must have the central, correct authority to resolve our problem". 

I will think more tomorrow and will let you know if I have cleared my doubts and mark the answer accepted. Huge, limitless thanks to you for bearing with me all along. <3 <3
